network/src/connection_manager.rs

```diff
@@ -0,0 +1,404 @@
+    /// 1. Attempt to connect to all default bootnodes on the network.
+    /// 2. Attempt to connect to all disconnected peers from the stored peer book.
+    ///
+    #[inline]
+    pub async fn new(
+        context: &Arc<Context>,
+        request_manager: RequestManager,
+        connections: Arc<RwLock<Connections>>,
+        storage: &Arc<MerkleTreeLedger>,
+        bootnode_addresses: Vec<SocketAddr>,
+        connection_frequency: u64,
+    ) -> Self {
+        // Load the preliminary local address from context.
+        let preliminary_local_address = context.local_address.read().await;
+
+        // Load the peer book from storage.
+        let peer_book = PeerBook::load(&storage).unwrap_or_else(|| {
+            // If the load fails, either the peer book does not exist,
+            // or it has been corrupted (e.g. from a data structure change).
+            let local_address = format!("0.0.0.0:{}", preliminary_local_address.port())
+                .parse()
+                .expect("local address");
+            let mut peer_book = PeerBook::new(local_address);
+            // Store the new peer book into the database.
+            peer_book
+                .store(&storage)
+                .unwrap_or_else(|error| debug!("Failed to store peer book into database {}", error));
+            peer_book
+        });
+        drop(preliminary_local_address);
+
+        // Instantiate a connection manager.
+        let connection_manager = Self {
+            peer_book: Arc::new(RwLock::new(peer_book)),
+            request_manager,
+            storage: storage.clone(),
+            channels: HashMap::new(),
+            ping_pong: context.pings.clone(),
+            bootnode_addresses,
+            minimum_peer_count: context.min_peers,
+            connection_frequency,
+
+            tmp_connections: connections,
+        };
+
+        // Initialize the connection manager.
+        debug!("Initializing the connection manager...");
+        {
+            // 1. Attempt to connect to all default bootnodes on the network.
+            connection_manager.connect_to_bootnodes().await;
+            // 2. Attempt to connect to all disconnected peers from the stored peer book.
+            if !context.is_bootnode {
+                // Only attempt the connection if the node is not a bootnode.
+                connection_manager.connect_to_all_disconnected_peers().await;
+            }
+        }
+        // Completed initializing connection manager.
+        debug!("Completed initializing connection manager.");
+
+        connection_manager
+    }
+
+    /// Returns the local address of the node.
+    #[inline]
+    pub async fn get_local_address(&self) -> SocketAddr {
+        // Acquire the peer book reader.
+        let peer_book = self.peer_book.read().await;
+        // Get the local address of the node.
+        let local_address = peer_book.local_address().clone();
+        // Drop the peer book reader.
+        drop(peer_book);
```

Comment:
locks don't need to be explicitly dropped if it's the end of the block they've been acquired in - they are dropped automatically; the same applies to a few other spots later in the diff